dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/core/GeneConstructor.java

```diff
@@ -0,0 +1,351 @@
+      locationEnd = location;
+    }
+
+    int strandNumber = convertStrand(strand);
+
+    val attributes = line[8];
+    val attributeMap = parseAttributes(attributes);
+
+    ObjectNode feature = DEFAULT.createObjectNode();
+    feature.put("seqname", seqname);
+    feature.put("source", source);
+    feature.put("type", type);
+    feature.put("locationStart", locationStart);
+    feature.put("locationEnd", locationEnd);
+    feature.put("strand", strandNumber);
+    for (val kv : attributeMap.entrySet()) {
+      feature.put(kv.getKey(), kv.getValue());
+    }
+    return feature;
+  }
+
+  private int convertStrand(char strand) {
+    if (strand == '+') {
+      return 1;
+    } else if (strand == '-') {
+      return -1;
+    } else {
+      return 0;
+    }
+  }
+
+  private String getName(String symbol) {
+    return nameMap().getOrDefault(symbol, "");
+  }
+
+  private ArrayNode getDescription(String id) {
+    return synMap().getOrDefault(id, DEFAULT.createArrayNode());
+  }
+
+  private static Map<String, String> parseAttributes(String attributes) {
+    val attributeMap = new HashMap<String, String>();
+    val tokens = SEMICOLON.split(attributes);
+    for (val token : tokens) {
+      String[] kv = UNKNOWN_WHITESPACE.split(QUOTES.matcher(token.trim()).replaceAll(""));
+      if (kv.length == 2) {
+        attributeMap.put(kv[0], kv[1]);
+      }
+    }
+    return attributeMap;
+  }
+
+  private ObjectNode constructGeneNode(ObjectNode data) {
+    val gene = DEFAULT.createObjectNode();
+    gene.put("_gene_id", asText(data, "gene_id"));
+    gene.put("symbol", asText(data, "gene_name"));
+    gene.put("biotype", asText(data, "gene_biotype"));
+    gene.put("chromosome", asText(data, "seqname"));
+    gene.put("strand", asText(data, "strand"));
+    gene.put("start", asInt(data, "locationStart"));
+    gene.put("end", asInt(data, "locationEnd"));
+
+    // Joining data from Ensembl and NCBI
+    gene.put("name", getName(asText(gene, "symbol")));
+    gene.put("description", summaryMap.getOrDefault(asText(gene, "_gene_id"), ""));
+    gene.put("synonyms", getDescription(asText(gene, "_gene_id")));
+    gene.put("canonical_transcript_id", canonicalMap().get(asText(gene, "_gene_id")));
+    return gene;
+  }
+
+  /**
+   * This method performs the calculations on transcript and exons to correctly mark coding regions. Heavily based on:
+   * https://github.com/icgc-dcc/dcc-heliotrope/blob/working/src/main/scripts/Heliotrope/Update/Ensembl.pm#L885
+   * 
+   * @param transcript ObjectNode representation of transcript with exons if any are present
+   * @param strand marks if + or - strand
+   * @return Processed Transcript ObjectNode
+   */
+  private ObjectNode postProcessTranscript(ObjectNode transcript, String strand, String canonical) {
+    if (asText(transcript, "id").equals(canonical)) {
+      transcript.put("is_canonical", true);
+    } else {
+      transcript.put("is_canonical", false);
+    }
```

Comment:
` transcript.put("is_canonical", isCanonical(transcript));` would reduce 4 lines from this already long-ish mainline.